django/db/backends/postgresql/base.py

```diff
@@ -272,7 +330,12 @@ def get_new_connection(self, conn_params):
                     f"Invalid transaction isolation level {isolation_level_value} "
                     f"specified. Use one of the psycopg.IsolationLevel values."
                 )
-        connection = self.Database.connect(**conn_params)
+        if self.pool:
+            # If nothing else has opened the pool, open it now
+            self.pool.open()
+            connection = self.pool.getconn()
+        else:
+            connection = self.Database.connect(**conn_params)
         if set_isolation_level:
             connection.isolation_level = self.isolation_level
```

Comment:
This line gave me pause; it modifies an attribute on a connection that (being part of a pool) is now reused. It seems like something better done in `configure_connection`. But, I guess this was already the case with persistent connections and if it's fine there, probably also fine here.

Reply:
Are you referring to `connection.isolation_level = self.isolation_level`? I believe that was done here 76356d963c131252e1ce4285083fd21fd0bdedd9 as part of [#24318](https://code.djangoproject.com/ticket/24318) if that helps 👍

Reply:
Ah sorry, I think I get what you mean now... true 🤔

Reply:
I added a test `test_connect_pool_with_isolation_level` to try and cover when it gets set "twice". I don't think it's doing any harm?

Reply:
From looking at the code (haven't verified in practice), I think the problem can happen in the following scenario:

- `isolation_level` is *not* configured in options (so `set_isolation_level` is false). In this case READ_COMMITTED is expected.
- User gets a connection from the pool.
- User sets `conn.isolation_level = SERIALIZABLE`
- User closes the connection (returns to the pool)
- User gets a connection from the pool (gets same reused connection from before)
- Expected: isolation level is READ_COMMITTED
- Actual: isolation level is SERIALIZABLE

I should note that the same issue happens with persistent connections, and I also see now that it is clearly documented as a caveat here https://docs.djangoproject.com/en/5.0/ref/databases/#caveats:

> If you modify parameters such as the connection’s isolation level or time zone, you should either restore Django’s defaults at the end of each request, force an appropriate value at the beginning of each request, or disable persistent connections.

So perhaps the fix is to have a similar note for the connection pool. That said, given that the pool always goes through the `DatabaseBackend` it *is* possible make this "just work" by making sure to never skip setting `isolation_level`, at least when the pool is used.

Reply:
> isolation_level is not configured in options (so set_isolation_level is false). In this case READ_COMMITTED is expected.

I think `connection.isolation_level` is None in this case as `set_isolation_level` is False and it will then take the Database default, which is probably `READ_COMMITTED`.

> User sets conn.isolation_level = SERIALIZABLE

How do they do this sorry?

Reply:
> How do they do this sorry?

Looking now, there does not seem to be a sanctioned way to do it from Django. And 3rd party libraries like https://github.com/Opus10/django-pgtransaction do it with an SQL statement, not setting the psycopg connection property. So it doesn't seems like a scenario that Django needs to concern itself with.

Reply:
Thank you for looking into it 🙇‍♀️

Reply:
I will be perfectly honest. I am not sure I 100% understand what the transaction isolation code is doing here and why it is written as it is. It should be possible to move this into `_configure_connection` as well. But maybe not since the base backend calls:
```
        self.connection = self.get_new_connection(conn_params)
        self.set_autocommit(self.settings_dict["AUTOCOMMIT"])
        self.init_connection_state()
```
so we would have to investigate whether changing the transcation level after the autocommit settings will work. In hindsight I now realize that we also change the timezone/role earlier when using the pool as opposed to the old behavior where it is done in `init_connection_state`. Can this cause problems? I think no, but…

Reply:
Thinking more about this, the comment above:
```
        # - before calling _set_autocommit() because if autocommit is on, that
        #   will set connection.isolation_level to ISOLATION_LEVEL_AUTOCOMMIT.
```
is no longer true (https://www.psycopg.org/docs/connection.html#connection.isolation_level):
```
This value is now entirely separate from the [autocommit](https://www.psycopg.org/docs/connection.html#connection.autocommit) property: in previous version, if autocommit was set to True this property would have returned [ISOLATION_LEVEL_AUTOCOMMIT](https://www.psycopg.org/docs/extensions.html#psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT); it will now return the server isolation level.
```
so we might be able to adjust this as needed. Then again we probably have to set this before setting autocommit to off because otherwise it would affect only the current transaction? Needs more thought.